mlir/include/mlir/Transforms/DialectConversion.h

```diff
@@ -1118,6 +1119,14 @@ struct ConversionConfig {
   // already been modified) and iterators into past IR state cannot be
   // represented at the moment.
   RewriterBase::Listener *listener = nullptr;
+
+  /// If set to "true", the dialect conversion driver attempts to fold
+  /// operations throughout the conversion. This is problematic because op
+  /// folders may assume that the IR is in a valid state at the beginning of
+  /// the folding process. However, the dialect conversion does not guarantee
+  /// that because some IR modifications are delayed until the end of the
+  /// conversion.
+  bool foldOps = true;
```

Comment:
I wonder whether there should be either a TODO such as "change to false in the future" (if we want to take that route) or whether the comment should note that it is true for historic reasons.

Looks funny that the majority of the paragraph discourages using the options but we default to it being true

Reply:
I would go further: if we consider this unsafe, we should just deprecate this mode entirely.

Reply:
Do mean setting this to `false` by default or adding a comment that this is deprecated (or both)? We have at least one test case in the test dialect that tests the folding.

Reply:
Yes we should be setting this to false by default, folks who are broken can set it back to true, but we also document it as deprecated.

Reply:
These tests are broken with `foldOps = false`:
```
Failed Tests (8):
  MLIR :: Conversion/AffineToStandard/lower-affine.mlir
  MLIR :: Conversion/FuncToLLVM/calling-convention.mlir
  MLIR :: Conversion/MemRefToLLVM/expand-then-convert-to-llvm.mlir
  MLIR :: Conversion/ShapeToStandard/shape-to-standard.mlir
  MLIR :: Conversion/VectorToLLVM/vector-to-llvm.mlir
  MLIR :: Dialect/LLVM/lower-to-llvm-e2e-with-target-tag.mlir
  MLIR :: Dialect/LLVM/lower-to-llvm-e2e-with-top-level-named-sequence.mlir
  MLIR :: Dialect/Vector/linearize.mlir
```

They are mostly FileCheck failures, but `vector-to-llvm.mlir` is actually broken. I'm busy with other stuff right now, so it might take a while. (Or if someone else wants to take this over, feel free to.)